### ISSUE TITLE
Multi agents interaction.

### ISSUE THREAD

**Issue by AGPX** [xbpeng/DeepMimic#156]
Hello,

congratulations, really a great job. I have a question: how hard is to have two opponents agents, in order to model an environment, for example, where an agent tries to get the soccer ball from his opponent (dribbling)? The ability to have 2 (or more) agents interacting can make this algorithm more suitable for a game.

**Comment by AGPX** [comment]
Maybe I made some progress (2 agents):

![image](https://user-images.githubusercontent.com/5240282/124193243-c77a6880-dac6-11eb-9db0-8f94616efd2c.png)

![image](https://user-images.githubusercontent.com/5240282/124193199-b7628900-dac6-11eb-8d2f-1b3fccf83eff.png)

one issue I had to solve is that the two agents are placed in the same position and so they intersect, but it seems to me that multi agent is effectively supported. Am I right?

**Comment by AGPX** [comment]
Improvements... now I've a kinematic character for each agent!

![image](https://user-images.githubusercontent.com/5240282/124299310-939e5200-db5d-11eb-95ad-217b470f9057.png)

**Comment by AGPX** [comment]
Mmhh.... looks like (after 5M of samples per agent) that Agent 1 (Spinkick) have some influences on Agent 0 (walk):

https://youtu.be/cbuAvkK0KG0

Still investigating...

**Comment by xbpeng** [comment]
Nice! Looks like you are making good progress on this. Unfortunately it will probably require pretty large changes to have the code support multiple characters. It will take a fair bit of work, but should be doable.

**Comment by AGPX** [comment]
Yes, I also believe it is doable (at least, to some extent on my part that I don't have a deep and complete understanding of your code). However, I made all my changes very quickly (in a couple of hours), but looking back at them I have already discovered some errors (read: lack of decoupling between agents) that most likely be the cause of bad training results. Stay tuned.

**Comment by AGPX** [comment]
Ok, it still doesn't work. I have solved many problems and now DeepMimicCore seems correct to me, but the training still does not work as it should. Now, my focus has shifted from C++ core to Python code. In particular, I think MPISolver may be the problem because it most likely mixes the data from Agent 0 and Agent 1. I definitely need 2 separate channels.

**Comment by AGPX** [comment]
Ok,

here is the result after separating the MPI communication channels for the agents:

https://youtu.be/m8ag-qKAasM

It seems that the walking agent learns to fall before hitting the other agent... I have to move it not only on x axis but also on z (actually we only have 'char_init_pos_xs' in the scene parameters, I have to add the equivalent for the z). Aside from that, it looks like training went well this time.

**Comment by AGPX** [comment]
I've displaced the actors to avoid collisions, but the walking agent (after 15M of samples) still falls at some point:

https://youtu.be/KlV2qKU4U18

Perhaps this behavior is due to the fact that the environment is always reset for both agents (that is, when one agent goes down, the other also gets reset, even if it was going well). Unfortunately, I don't think there's a work-around for this. Anyway, I try to keep the learning going, but I'm afraid Agent 0 won't recover easily...

**Comment by xbpeng** [comment]
The intermediate results look promising. 15M samples is not a lot, it usually takes 50-100m samples to learn a motion. So I think if you keep training this, and it should still get better.

You can probably modify the reset so that it only resets the agent that falls rather than both agents.

**Comment by AGPX** [comment]
> You can probably modify the reset so that it only resets the agent that falls rather than both agents.

This looks very difficult to implement, especially since the reset usually apply to the world as well (which includes shared things, like forces) and I don't really know how to apply it to a single, isolated, agent.
I'm starting to think that it would have been easier to have multiple cScene objects (rather than multiple actors), sharing the same cWorld and cGround instances (and also the perturbations), but the problem of the separate reset remains.
